### PR TITLE
Severity-based logging -- parent PR

### PR DESCRIPTION
**Pull request by jonatack**
This is a parent PR for continuing updates to severity-based logging. See https://github.com/bitcoin/bitcoin/issues/20576 for motivation and https://github.com/bitcoin/bitcoin/pull/25306 for discussion.

- [x] https://github.com/bitcoin/bitcoin/pull/24464
- [x] https://github.com/bitcoin/bitcoin/pull/25286
- [x] #25306
- [x] https://github.com/bitcoin/bitcoin/issues/25614
- [x] https://github.com/bitcoin/bitcoin/pull/26957
- [x] https://github.com/bitcoin/bitcoin/pull/27632
- #27231 
- print the wallet name more clearly in wallet logging to distinguish it from category/level
- unconditionally log `Info` severity level messages
- replace the hardcoded `LogLevelsList()` vector with a programmatic one derived from the `Level` enum class
- convert `GetLogCategory()` to `std::optional`
- deduplicate the `LogCategory` code
- update to severity-based logging in the following areas, dropping the use of `LogPrintf` and `LogPrint`: addrdb, addrman, banman, i2p, mempool, netbase, net, net_processing, timedata, torcontrol
- when we're ready to drop the other logging macros, rename `LogPrintLevel` to simply `Log`, which will be the only log macro needed after migration from the other macros is complete

### CODE REVIEW

File: src/i2p.cpp [hunk 1]
```diff
@@ -376,7 +376,7 @@ void Session::CreateIfNotCreatedAlready()
     m_session_id = session_id;
     m_control_sock = std::move(sock);
 
-    LogPrintf("I2P: SAM session created: session id=%s, my address=%s\n", m_session_id,
+    LogPrintf("[i2p] SAM session created: session id=%s, my address=%s\n", m_session_id,
```
Comment (laanwj) [inline src/i2p.cpp:379]: I understand the idea, but I don't think we should do it this way by inserting `[cat]` manually. We need a way to log things with a category even without `-debug=`.
This is why I have a commit in #25202 to log everything >= Warning unconditionally. I'm not sure that is the best solution but I like it more than this.

Comment (jonatack) [reply src/i2p.cpp:379]: Thanks, SGTM, will just do the de-dupe then (also for ZMQ and PRUNE).

Comment (laanwj) [reply src/i2p.cpp:379]: That's fine but to be clear I mean I agree conceptually with doing it, just not in this way.

Comment (jonatack) [reply src/i2p.cpp]: Done.

Comment (jonatack) [reply src/i2p.cpp]: > That's fine but to be clear I mean I agree conceptually with doing it, just not in this way.

Agree, thanks for clarifying.  I'll start with reviewing your change and then go from there / update here.

Comment (jonatack) [reply src/i2p.cpp]: How about `LogPrintfCategory`?
```diff
--- a/src/logging.h
+++ b/src/logging.h
@@ -198,6 +198,7 @@ static inline void LogPrintf_(const std::string& logging_function, const std::st
 
 #define LogPrintf(...) LogPrintLevel_(BCLog::LogFlags::NONE, BCLog::Level::None, __VA_ARGS__)
+#define LogPrintfCategory(category, ...) LogPrintLevel_(category, BCLog::Level::None, __VA_ARGS__)

--- a/src/i2p.cpp
+++ b/src/i2p.cpp
@@ -376,7 +376,7 @@ void Session::CreateIfNotCreatedAlready()
 
-    LogPrintf("I2P: SAM session created: session id=%s, my address=%s\n", m_session_id,
+    LogPrintfCategory(BCLog::I2P, "SAM session created: session id=%s, my address=%s\n", m_session_id,
```

debug log
```
2022-05-25T16:03:02Z [i2p] SAM session created
```

Comment (jonatack) [reply src/i2p.cpp]: Done

**Comment by jonatack** [comment]
This will need to be updated after the merge of #25202 for the change in [refactor: Change LogPrintLevel order to category, severity](https://github.com/bitcoin/bitcoin/pull/25202/commits/c4e77177276ea2b79c4675cb2678ee2cc757b743), so leaving in draft for now.

File: src/logging.h [hunk 2]
```diff
@@ -203,6 +203,7 @@ static inline void LogPrintf_(const std::string& logging_function, const std::st
 #define LogPrintLevel_(category, level, ...) LogPrintf_(__func__, __FILE__, __LINE__, category, level, __VA_ARGS__)
 
 #define LogPrintf(...) LogPrintLevel_(BCLog::LogFlags::NONE, BCLog::Level::None, __VA_ARGS__)
+#define LogPrintfCategory(category, ...) LogPrintLevel_(category, BCLog::Level::None, __VA_ARGS__)
```
Comment (MarcoFalke) [inline src/logging.h:206]: Would be good to add a docstring for this. Is this meant to unconditionally log regardless of log level? If yes, it doesn't seem it would work. `BCLog::Level::None` is `1` but the check for unconditional logging is `>= Warning (3)`?

Comment (jonatack) [reply src/logging.h:206]: Good catch; this is wrong for `LogPrintf` as well. Likely going to fix by reordering the `BCLog::Level` enums:
```diff
diff --git a/src/logging.cpp b/src/logging.cpp
index f1a86f0dce..3f316f0101 100644
--- a/src/logging.cpp
+++ b/src/logging.cpp
@@ -187,8 +187,6 @@ bool GetLogCategory(BCLog::LogFlags& flag, const std::string& str)
 std::string LogLevelToStr(BCLog::Level level)
 {
     switch (level) {
-    case BCLog::Level::None:
-        return "none";
     case BCLog::Level::Debug:
         return "debug";
     case BCLog::Level::Info:
@@ -197,6 +195,8 @@ std::string LogLevelToStr(BCLog::Level level)
         return "warning";
     case BCLog::Level::Error:
         return "error";
+    case BCLog::Level::None:
+        return "";
     }
     assert(false);
 }
```
```diff
diff --git a/src/logging.h b/src/logging.h
index 8a896b6b33..f8e55b88b2 100644
--- a/src/logging.h
+++ b/src/logging.h
@@ -69,10 +69,10 @@ namespace BCLog {
     };
     enum class Level {
         Debug = 0,
-        None = 1,
-        Info = 2,
-        Warning = 3,
-        Error = 4,
+        Info,
+        Warning,
+        Error,
+        None,
     };
```

Comment (jonatack) [reply src/logging.h:206]: >  it doesn't seem it would work. BCLog::Level::None is 1 but the check for unconditional logging is >= Warning (3)?

Ah, `LogAcceptCategory()` with the logic you mention is called by neither `LogPrintf` nor `LogPrintfCategory`, so it seems fine as-is. Will add a docstring.

Comment (jonatack) [reply src/logging.h:209]: > Would be good to add a docstring for this.

Thanks for the suggestion; done.

File: src/i2p.cpp (hunk 1 above)
Comment (laanwj) [reply src/i2p.cpp]: Yes, lgtm!

File: src/logging.h [hunk 3]
```diff
@@ -199,11 +200,14 @@ static inline void LogPrintf_(const std::string& logging_function, const std::st
...
 }
 
-
 #define LogPrintLevel_(category, level, ...) LogPrintf_(__func__, __FILE__, __LINE__, category, level, __VA_ARGS__)
 
+// Log unconditionally.
 #define LogPrintf(...) LogPrintLevel_(BCLog::LogFlags::NONE, BCLog::Level::None, __VA_ARGS__)
 
+// Log unconditionally, prefixing the output with the passed category name.
+#define LogPrintfCategory(category, ...) LogPrintLevel_(category, BCLog::Level::None, __VA_ARGS__)
```
Comment (laanwj) [inline src/logging.h:209]: It's `LogPrintLevel` and `LogPrintfCategory` that's a bit inconsistent. I think we can leave out the `f` here.

**Comment by klementtan** [comment]
Approach ACK

**Comment by laanwj** [comment]
Concept ACK

**Comment by jonatack** [comment]
Rebased/updated following the merge of #25286 that contained the first commit of this PR.

File: src/logging.h [hunk 4]
```diff
@@ -68,11 +68,12 @@ namespace BCLog {
         ALL         = ~(uint32_t)0,
     };
     enum class Level {
-        Debug = 0,
-        None = 1,
-        Info = 2,
-        Warning = 3,
-        Error = 4,
+        Trace = 0,
```
Comment (dergoegge) [inline src/logging.h:71]: I think some docs on the different levels wouldn't hurt. For example, it's not quite clear to me how you choose between `Trace` and `Debug`.

Comment (laanwj) [reply src/logging.h:71]: Generally, the distinction is:
- debug is for reasonably noisy logging, but not high enough volume to make it unusable in a production scenario
- tacing is super high detail, step by step, and will slow down the application and you'd definitely never want to enable it for a category unless you're developing or trouble-shooting a specific issue

**Review by Onyeali — APPROVED** [review]

**Comment by jonatack** [comment]
Rebased and updated significantly to pull in the changes in #25287 (with review feedback and some clean-up and moving the refactoring commits to after the functional ones) and the changes discussed in #25306 (unconditionally log Info level and up). Will re-update the specific log messages themselves after the first commits here are confirmed or merged.

File: src/logging.h (hunk 4 above)
Comment (jonatack) [reply src/logging.h]: Thanks! Added documentation to the Level enum class.

**Comment by jonatack** [comment]
Rebased and updated with the latest changes in subset PR #25614.

**Review by w0xlt — COMMENTED** [review]
Approach ACK.

This PR improves significantly the quality of the log and the Log codebase.
I wonder if the main Log classes could be a separate library in another repository.
Awaiting rebase for ACK.

**Comment by vasild** [comment]
Approach ACK

**Comment by jonatack** [comment]
Up for grabs.

**Comment by MarcoFalke** [comment]
Concept ACK

**Comment by jonatack** [comment]
A big thank you to @vasild who rebased this branch to current master. Re-opening and updating.

File: src/net_processing.cpp [hunk 5]
```diff
@@ -2356,9 +2356,9 @@ void PeerManagerImpl::HeadersDirectFetchBlocks(CNode& pfrom, const Peer& peer, c
         // the main chain -- this shouldn't really happen.  Bail out on the
         // direct fetch and rely on parallel download instead.
         if (!m_chainman.ActiveChain().Contains(pindexWalk)) {
-            LogPrint(BCLog::NET, "Large reorg, won't direct fetch to %s (%d)\n",
-                    pindexLast->GetBlockHash().ToString(),
-                    pindexLast->nHeight);
+            LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "Large reorg, won't direct fetch to %s (%d)\n",
```
Comment (dergoegge) [inline src/net_processing.cpp:2359]: Is the switch to unconditional logging intended here?

If yes, then it's probably better done in a separate commit because it is not immediately clear that this is safe and it's quite easy to miss mixed in with all the other changes in this commit.

File: src/logging.cpp [hunk 6]
```diff
@@ -125,9 +125,9 @@ bool BCLog::Logger::WillLogCategory(BCLog::LogFlags category) const
 
 bool BCLog::Logger::WillLogCategoryLevel(BCLog::LogFlags category, BCLog::Level level) const
 {
-    // Log messages at Warning and Error level unconditionally, so that
-    // important troubleshooting information doesn't get lost.
-    if (level >= BCLog::Level::Warning) return true;
+    // Log the Info, Warning, and Error message levels unconditionally, so that
+    // important troubleshooting information isn't lost.
+    if (level >= BCLog::Level::Info) return true;
```
Comment (dergoegge) [inline src/logging.cpp:131]: I feel like this almost deserves its own PR (maybe you are already planning on doing so). afaict we don't log anything right now using `Info` which is the only reason this change is safe but that might not be clear to reviewers of this rather large PR. (If for example we had a lot of log locations using `Info` then they would all need to be revisited w.r.t. to disk filling attacks)

**Comment by jonatack** [comment]
Updating and rebasing.

**Comment by jonatack** [comment]
Thanks for the detailed feedback, @ajtowns. Will update with it in mind; looks like it will be helpful in simplifying things.

File: src/net_processing.cpp (hunk 5 above)
Comment (jonatack) [reply src/net_processing.cpp:2359]: Yes, good point that any changes from conditional to unconditional logging should be separated out.

File: src/logging.cpp (hunk 6 above)
Comment (jonatack) [reply src/logging.cpp:131]: Good point. My original thinking was that `info` would be conditional like `debug` and `trace` and that only `warning` and `error` would be unconditional, which would also reduce the volume of default logging. However, that changed during the discussions in https://github.com/bitcoin/bitcoin/pull/25306. Will keep your suggestion in mind while restructuring the update here.
